ryft-ui/src/main/java/com/ryftpay/android/ui/service/request/GooglePayRequestFactory.kt

```diff
@@ -0,0 +1,130 @@
+package com.ryftpay.android.ui.service.request
+
+import com.google.android.gms.wallet.IsReadyToPayRequest
+import com.google.android.gms.wallet.PaymentDataRequest
+import com.ryftpay.android.ui.extension.formatPriceWithoutCurrencySymbol
+import com.ryftpay.android.ui.model.RyftCardType
+import org.json.JSONArray
+import org.json.JSONObject
+import java.util.Currency
+
+internal object GooglePayRequestFactory {
+
+    private const val MAJOR_API_VERSION = 2
+    private const val MINOR_API_VERSION = 0
+    private const val CARD_PAYMENT_METHOD = "CARD"
+    private const val IS_EXISTING_PAYMENT_METHOD_REQUIRED = true
+    private const val PAYMENT_GATEWAY_TOKENIZATION_SPECIFICATION_TYPE = "PAYMENT_GATEWAY"
+    private const val RYFT_PAYMENT_GATEWAY = "ryft"
+    private val supportedAuthMethods = arrayOf("PAN_ONLY", "CRYPTOGRAM_3DS")
+    private val baseRequest: JSONObject = JSONObject()
+        .put("apiVersion", MAJOR_API_VERSION)
+        .put("apiVersionMinor", MINOR_API_VERSION)
+    private val cardPaymentMethod: JSONObject = JSONObject()
+        .put("type", CARD_PAYMENT_METHOD)
+        .put(
+            "parameters",
+            JSONObject()
+                .put("allowedAuthMethods", JSONArray(supportedAuthMethods))
+                .put(
+                    "allowedCardNetworks",
+                    JSONArray(
+                        RyftCardType.getGooglePaySupportedTypeNames()
+                    )
+                )
+        )
+
+    internal val isReadyToPayRequest: IsReadyToPayRequest = IsReadyToPayRequest.fromJson(
+        JSONObject(baseRequest.toString()).apply {
+            put("existingPaymentMethodRequired", IS_EXISTING_PAYMENT_METHOD_REQUIRED)
+            put("allowedPaymentMethods", JSONArray().put(cardPaymentMethod))
+        }.toString()
+    )
+
+    internal fun createPaymentDataRequest(
+        loadPaymentDataRequest: LoadPaymentDataRequest
+    ): PaymentDataRequest = PaymentDataRequest.fromJson(
+        JSONObject(baseRequest.toString()).apply {
+            put("merchantInfo", createMerchantInfo(loadPaymentDataRequest.merchantInfo))
+            put(
+                "allowedPaymentMethods",
+                JSONArray().put(
+                    createCardPaymentMethodWithTokenizationSpecification(
+                        loadPaymentDataRequest.tokenizationSpecification
+                    )
+                )
+            )
+            put(
+                "transactionInfo",
+                createTransactionInfo(loadPaymentDataRequest.transactionInfo)
+            )
+        }.toString()
+    )
+
+    private fun createMerchantInfo(
+        merchantInfo: MerchantInfo
+    ): JSONObject = JSONObject()
+        .put("merchantName", merchantInfo.displayName)
+
+    private fun createTokenizationSpecification(
```

Comment:
Might be worth exposing this publicy should any client want to place the GooglePay button on a product page/outside of checkout and handle the GooglePay interaction themselves.

We could add this in the next iteration however, with an example in the demo app.

Reply:
yeah good shout, I'll pick this up separately 👍